docs/source/tutorials/qat_finetune.rst

```diff
@@ -0,0 +1,371 @@
+    (output_proj): Linear(in_features=4096, out_features=4096, bias=False)
+    (pos_embeddings): RotaryPositionalEmbeddings()
+  )
+
+  >>> print(prepared_model.layers[0].attn)
+  CausalSelfAttention(
+    (q_proj): Int8DynActInt4WeightQATLinear(in_features=4096, out_features=4096, bias=False)
+    (k_proj): Int8DynActInt4WeightQATLinear(in_features=4096, out_features=1024, bias=False)
+    (v_proj): Int8DynActInt4WeightQATLinear(in_features=4096, out_features=1024, bias=False)
+    (output_proj): Int8DynActInt4WeightQATLinear(in_features=4096, out_features=4096, bias=False)
+    (pos_embeddings): RotaryPositionalEmbeddings()
+  )
+
+After fine-tuning, we can convert the model to get an actual quantized model.
+If we print the converted model, we’ll see that the QAT linears have been
+swapped with :code:`Int8DynActInt4WeightLinear`, which are the quantized versions
+of the linear layers. This quantized model can then be saved to checkpoint and
+used for inference or generation.
+
+.. code-block:: python
+
+  # Fine-tune as before
+  train_loop(prepared_model)
+
+  # Convert fake quantize to actual quantize operations
+  converted_model = quantizer.convert(prepared_model)
+
+.. code-block:: bash
+
+  >>> print(converted_model.layers[0].attn)
+  CausalSelfAttention(
+    (q_proj): Int8DynActInt4WeightLinear()
+    (k_proj): Int8DynActInt4WeightLinear()
+    (v_proj): Int8DynActInt4WeightLinear()
+    (output_proj): Int8DynActInt4WeightLinear()
+    (pos_embeddings): RotaryPositionalEmbeddings()
+  )
+
+
+QAT finetuning recipe in torchtune
+----------------------------------
+
+Putting it all together, we can now fine-tune a model using torchtune’s `QAT recipe <https://github.com/pytorch/torchtune/blob/main/recipes/qat_distributed.py>`_.
+Make sure that you have first downloaded the Llama3 weights and tokenizer by
+following :ref:`these instructions<download_llama_label>`. In this tutorial,
+we use the following settings to demonstrate QAT’s effectiveness in recovering
+quantization degradation compared to directly quantizing a model fine-tuned
+without QAT. You can copy the default QAT config and make the following
+modifications accordingly:
+
+.. code-block:: bash
+
+  tune cp llama3/8B_qat_full custom_8B_qat_full.yaml
+
+.. code-block:: yaml
+
+  # Dataset
+  dataset:
+    _component_: torchtune.datasets.text_completion_dataset
+    source: allenai/c4
+    max_seq_len: 8192
+    column: text
+    name: en
+    split: train
+  seed: null
+  shuffle: True
+
+  ...
+
+  epochs: 1
+  max_steps_per_epoch: 2000
+  fake_quant_after_n_steps: 1000
+  memory_efficient_fsdp_wrap: False
+
+.. note::
+
+  QAT in torchtune is currently not compatible with :code:`memory_efficient_fsdp_wrap`. This is a known issue and will be fixed in a future torchtune version.
+
+You can then use the following command to run fine-tuning with QAT using the above
+config. This workload requires at least 6 GPUs, each with VRAM of at least 80GB.
```

Comment:
Are there other techniques we can use to get below this? I know you mention below that `memory_efficient_fsdp_wrap` isn't supported, but what about low-precision optimizers or fusing optimizer and backward? If we can use these (or other memory-saving techniques) it might be a nice way to make the recipe accessible to a broader set of users. 

I think you mentioned previously that QAT typically uses about 30% more memory than a regular full-finetune, right? Might be worth at least briefly mentioning that the params need to get copied and so the expected memory usage for QAT will be higher than a regular full fine-tune (just to preempt any possible questions on that front).

Reply:
Thanks, added a note on QAT overheads